File: update_page.py
#!/usr/bin/env python3
"""
Fully-automated updater for the Horse Checker "Jobs in Racing" page.

It fetches the live Careers in Racing RSS feed, builds the job listings as
static HTML (your existing article.job-item markup) and writes them straight
into your real .html page — in place, between two marker comments. The page
stays a genuine static HTML file (great for SEO), it just gets refreshed
automatically.

Because it rebuilds the whole block from the current feed every run, jobs that
have dropped out of the feed disappear from the page automatically — no stale
listings to clean up by hand.

ONE-TIME SETUP
--------------
In your jobs page, wrap the listings area with these two comment markers
(put them inside <div class="job-listings"> ... </div>):

    <div class="job-listings">
    <!-- JOBS:START -->
    <!-- JOBS:END -->
    </div>

Anything between the markers is what this script replaces.

RUN IT
------
    python3 update_page.py /path/to/jobs-in-racing.html

Then schedule it (see cron.txt) and it runs on its own — zero clicks.
"""

import sys

from generate import build_articles, fetch_feed, FEED_URL

START = "<!-- JOBS:START -->"
END = "<!-- JOBS:END -->"


def main(argv):
    if len(argv) < 2:
        sys.exit("Usage: python3 update_page.py /path/to/jobs-in-racing.html [feed_url]")

    page_path = argv[1]
    feed = argv[2] if len(argv) > 2 else FEED_URL

    articles = build_articles(fetch_feed(feed))
    block = "\n".join(articles)

    with open(page_path, "r", encoding="utf-8") as fh:
        html = fh.read()

    if START not in html or END not in html:
        sys.exit(
            "Markers not found. Add these once inside your listings div:\n"
            f"  {START}\n  {END}"
        )

    before, rest = html.split(START, 1)
    _, after = rest.split(END, 1)
    new_html = f"{before}{START}\n{block}\n    {END}{after}"

    with open(page_path, "w", encoding="utf-8") as fh:
        fh.write(new_html)

    print(f"Updated {page_path} with {len(articles)} jobs.")


if __name__ == "__main__":
    main(sys.argv)

File: update_page.php
<?php
/**
 * update_page.php — fully-automatic updater for the Horse Checker jobs page.
 *
 * Run by a cron job. It fetches the Careers in Racing RSS feed and writes the
 * job listings straight into your existing .html page, in place, between two
 * marker comments. Your page stays a normal static .html file — no renaming,
 * no changing any links, and search engines still see real HTML.
 *
 * Rebuilds from the current feed every run, so jobs that leave the feed drop
 * off the page automatically.
 *
 * USAGE (command line / cron):
 *   php update_page.php /full/path/to/jobs-in-racing.html
 *
 * ONE-TIME: inside your <div class="job-listings"> put these two markers
 * (delete your current manual listings between them):
 *   <!-- JOBS:START -->
 *   <!-- JOBS:END -->
 */

$FEED_URL = 'https://jobs.careersinracing.com/jobsrss/?Sector=1&countrycode=GB';
$START = '<!-- JOBS:START -->';
$END   = '<!-- JOBS:END -->';

if ($argc < 2) {
    fwrite(STDERR, "Usage: php update_page.php /path/to/jobs-in-racing.html [feed_url]\n");
    exit(1);
}
$page = $argv[1];
$feed = $argc > 2 ? $argv[2] : $FEED_URL;

function hc_clean($s) { return trim(preg_replace('/\s+/', ' ', (string)$s)); }

$ctx = stream_context_create(['http' => [
    'timeout' => 20,
    'header'  => "User-Agent: HorseChecker-Jobs/1.0\r\n",
]]);
$xml = @file_get_contents($feed, false, $ctx);
if ($xml === false) { fwrite(STDERR, "Could not fetch feed: $feed\n"); exit(1); }

$rss = @simplexml_load_string($xml);
if (!$rss || !isset($rss->channel->item)) { fwrite(STDERR, "No items in feed.\n"); exit(1); }

$today = date('Y-m-d');
$blocks = [];
foreach ($rss->channel->item as $item) {
    $title = hc_clean($item->title);
    $desc  = hc_clean($item->description);
    $link  = hc_clean($item->link);
    if ($title === '' && $desc === '' && $link === '') continue;

    $ts     = strtotime((string)$item->pubDate) ?: time();
    $iso    = gmdate('Y-m-d\TH:i:s\Z', $ts);
    $posted = 'Posted: ' . date('l, d M Y', $ts);
    $isNew  = (date('Y-m-d', $ts) === $today) ? 'New Today' : '';

    $L = htmlspecialchars($link, ENT_QUOTES);
    $T = htmlspecialchars($title, ENT_QUOTES);
    $D = htmlspecialchars($desc, ENT_QUOTES);
    $P = htmlspecialchars($posted, ENT_QUOTES);
    $blocks[] = <<<HTML
  <article class="job-item">
    <h2><a href="$L" target="_blank" rel="nofollow">$T</a></h2>
    <p class="job-description">$D</p>
    <p class="new">$isNew</p>
    <a href="$L" target="_blank" rel="nofollow" class="apply-link"><button type="button" class="custom-btn">Apply Here</button></a>
    <time datetime="$iso">$P</time>
  </article>
HTML;
}
$block = implode("\n", $blocks);

$html = @file_get_contents($page);
if ($html === false) { fwrite(STDERR, "Cannot read page: $page\n"); exit(1); }

$sPos = strpos($html, $START);
$ePos = strpos($html, $END);
if ($sPos === false || $ePos === false || $ePos < $sPos) {
    fwrite(STDERR, "Markers not found. Add these once inside your listings div:\n  $START\n  $END\n");
    exit(1);
}

$before = substr($html, 0, $sPos);
$after  = substr($html, $ePos + strlen($END));
$new    = $before . $START . "\n" . $block . "\n    " . $END . $after;

if (@file_put_contents($page, $new) === false) {
    fwrite(STDERR, "Cannot write page: $page (check file permissions)\n");
    exit(1);
}
echo "Updated $page with " . count($blocks) . " jobs.\n";
